### PR TITLE
Mesh Untangling and worst case quality improvement metrics (TMOP)

### PR DESCRIPTION
**Pull request by kmittal2**
This PR adds non-barrier metrics that can be used for mesh untangling and worst case quality improvement using TMOP.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3000](https://github.com/mfem/mfem/pull/3000) | @kmittal2 | @tzanio | @vladotomov + @tzanio | 6/4/22 | 7/13/22 | 7/24/22 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: miniapps/meshing/mesh-optimizer.cpp [hunk 1]
```diff
@@ -100,6 +100,10 @@
 //
 //   2D untangling:
 //     mesh-optimizer -m jagged.mesh -o 2 -mid 22 -tid 1 -ni 50 -li 50 -qo 4 -fd -vl 1
+//   2D untangling with shifted barrier metric:
+//     mesh-optimizer -m jagged.mesh -o 2 -mid 4 -tid 1 -ni 500 -li 50 -qo 4 -fd -vl 1 -um 1 -no-shifted
```
Comment (vladotomov) [inline miniapps/meshing/mesh-optimizer.cpp:104]: Update options.

File: miniapps/meshing/mesh-optimizer.cpp [hunk 2]
```diff
@@ -35,7 +35,7 @@
 //   Adapted analytic shape:
 //     mesh-optimizer -m square01.mesh -o 2 -rs 2 -mid 2 -tid 4 -ni 200 -bnd -qt 1 -qo 8
 //   Adapted analytic size+orientation:
-//     mesh-optimizer -m square01.mesh -o 2 -rs 2 -mid 14 -tid 4 -ni 100 -bnd -qt 1 -qo 8 -fd
+//     mesh-optimizer -m square01.mesh -o 2 -rs 2 -mid 14 -tid 4 -ni 200 -bnd -qt 1 -qo 8 -fd
```
Comment (vladotomov) [inline miniapps/meshing/mesh-optimizer.cpp:38]: revert

File: fem/tmop_tools.hpp [hunk 3]
```diff
@@ -136,6 +136,7 @@ class TMOPNewtonSolver : public LBFGSSolver
 
    // Minimum determinant over the whole mesh. Used for mesh untangling.
    double *min_det_ptr = nullptr;
+   mutable bool compute_metric_quantile_flag = true;
```
Comment (vladotomov) [inline fem/tmop_tools.hpp:139]: Comment what it's used for.

File: fem/tmop_tools.cpp [hunk 4]
```diff
@@ -437,6 +437,7 @@ double TMOPNewtonSolver::ComputeScalingFactor(const Vector &x,
    const double norm_in = Norm(r);
 
    const double detJ_factor = (solver_type == 1) ? 0.25 : 0.5;
+   compute_metric_quantile_flag = false;
```
Comment (vladotomov) [inline fem/tmop_tools.cpp:440]: Shouldn't these be updated during the line search? They affect the energy and the residual.

File: fem/tmop.hpp [hunk 5]
```diff
@@ -104,6 +104,83 @@ class TMOP_Combo_QualityMetric : public TMOP_QualityMetric
...
+                                          int exponent_ = 1,
+                                          double alpha_ = 1.5,
+                                          double detT_ep_ = 0.0001,
+                                          double muT_ep_ = 0.0001,
+                                          BarrierType btype_ = BarrierType::None,
+                                          WorstCaseType wctype_ = WorstCaseType::None) :
+      tmop_metric(tmop_metric_), exponent(exponent_), alpha(alpha_),
+      detT_ep(detT_ep_), muT_ep(muT_ep_), btype(btype_), wctype(wctype_) { }
+
+   virtual void SetTargetJacobian(const DenseMatrix &Jtr_)
```
Comment (vladotomov) [inline fem/tmop.hpp:159]: not needed

File: fem/tmop.hpp [hunk 6]
```diff
@@ -1523,6 +1638,11 @@ class TMOP_Integrator : public NonlinearFormIntegrator
    void AssemblePA_Limiting();
    void ComputeAllElementTargets(const Vector &xe = Vector()) const;
 
+   double ComputeUntanglerMinDetT(const Vector &x,
+                                  const FiniteElementSpace &fes);
+   double ComputeUntanglerMaxMuT(const Vector &x,
```
Comment (vladotomov) [inline fem/tmop.hpp:1643]: I thought this was not necessarily untangle-related. How about `ComputeMaxMetricValue`?

**Review by vladotomov — APPROVED** [review]

File: fem/tmop.hpp [hunk 7]
```diff
@@ -104,6 +104,89 @@ class TMOP_Combo_QualityMetric : public TMOP_QualityMetric
...
+   };
+
+protected:
+   TMOP_QualityMetric &tmop_metric; //non-barrier metric to use
+   double min_detT;                 //minimum Jacobian in the mesh
+   double max_muT;                  //max mu_k/phi(tau,ep) in the mesh
+   int exponent;                    //used for p-mean metrics
+   double alpha;                    //scaling factor for min(det(T))
+   double detT_ep;                  //small constant subtracted from min(detT)
+   double muT_ep;                   //small constant added to muT term
```
Comment (tzanio) [inline fem/tmop.hpp:144]:
```suggestion
   TMOP_QualityMetric &tmop_metric; // non-barrier metric to use
   double min_detT;                 // minimum Jacobian in the mesh
   double max_muT;                  // max mu_k/phi(tau,ep) in the mesh
   int exponent;                    // used for p-mean metrics
   double alpha;                    // scaling factor for min(det(T))
   double detT_ep;                  // small constant subtracted from min(detT)
   double muT_ep;                   // small constant added to muT term
```

File: fem/tmop_tools.cpp [hunk 8]
```diff
@@ -437,6 +437,9 @@ double TMOPNewtonSolver::ComputeScalingFactor(const Vector &x,
...
+   // TODO customized line search for worst-quality optimization.
+   // TODO what is the Newton exit criterion for worst-quality optimization.
```
Comment (tzanio) [inline fem/tmop_tools.cpp:442]:
```suggestion
   // TODO:
   // - Customized line search for worst-quality optimization.
   // - What is the Newton exit criterion for worst-quality optimization?
```

**Review by tzanio — APPROVED** [review]
Nice work, @kmittal2 !

**Pull request merged**